### ISSUE TITLE
Ingest CNCF vids

### ISSUE THREAD

**Issue by erkannt** [sciety/sciety#663]
## TLDR

- k8s model of orchestration, CRDs and operators is lending itself nicely to building solutions for serverless, edge compute and infrastructure management.
- GitOps leads one to identifying areas of applications and their operation that can't be modelled declaratively, managed by idempotent operations or require a sequence of steps e.g. DB schema migrations.
- The evolution of e.g. Ansible -> Terraform -> Crossplane to me reflects an understanding of the power of operators and our ability to go from imperative to declarative interaction with machines.
- _Serverless_ doesn't only mean Faas. I like the term _nodeless_ and the phrase _hiding the machines_.
- Alternatives to yaml are emerging.
- Crossplane went  1.0 in December. All major cloud vendors are building tools to manage your infra as CRDs using k8s.
- k8s (just like most runtimes/operating systems) has interesting failure modes when you push it to various limits.
- The evolution of Helm's chart collection suggests interesting parallels for Sciety. I also wonder if Sciety and the groups it harbours are the _complement_ to science and their funding should be seen the same way that e.g. Intel sees its investment into Linux.
- Managing and linking events between cloud services declaratively is an interesting space with lots of current activity.
- k3s and k3sup look really nice for a) running k8s locally (sadly only Linux native, but can be run in Docker, WSL support in progress) and b) testing and developing things like flux at lower cost than full EKS + workers.

### [K3s Under the Hood: Building a Product-grade Lightweight Kubernetes Distro - Darren Shepherd](https://www.youtube.com/watch?v=-HchRyqNtkU)

Interesting overview but also let me understand k8s more by learning about what is different in k3s.

- 150mb for worker only, 300mb with control plane
- intended for prod use, not only dev work
- single node 'cluster', single server multiple agents, multiple servers multiple agents, internal or external db for controlplane
- has something like `--watch` that you can point at a directory with manifests
- includes traefik for ingress
- has been scaled to 1000s of nodes
- `k3sup` for deploying k3s
- Civo offers k3s as a managed service

![image](https://user-images.githubusercontent.com/19282025/115865473-46e91d00-a430-11eb-8a65-fb4c98f5cfb7.png)

### [A Flight Over the Cloud Native Landscape - Carson Anderson, Weave](https://www.youtube.com/watch?v=icvjuNCYi7c)

Nice rapid run through.

- what does CNCF graduated mean? no measure of production readiness, but at least two orgs contribute
- Contour (cloud native ingress controller)
- KubeEdge (use k8s to manage edge compute)
- Rook (storage orchestration)
- presentation done with inkscape and sozi.baierouge.fr

### [Helm: Past, Present, Future - Matt Butcher & Bridget Kromhout, Microsoft, & Matt Farina](https://www.youtube.com/watch?v=thaJ3k3ZhBY)

Good for understanding the evolution of helm. Their transition from growing and maintaining a central repo of chart to decentralising it is an interesting thought for Sciety.

- helm org automated review of 'charts' to increase the content available to the 'package manager'
- still loosing sleep over the choice of templating lang (but would still make that choice)
- background on why they went from `stable` to decentral charts and artifact hub
- CRDs in helm are still really hard because they could break deployments outside of the chart
- Helm 3 support charts coming from OCI-base registries https://helm.sh/docs/topics/registries/

### [End User Panel: GITOPS in the Enterprise -Real World Experiences - Cheryl Hung](https://www.youtube.com/watch?v=qlbVMj1L9dY)

Made me reflect on our use of fluxcd and the test/release of reviewer.

- reasons and benefits:
  - disaster recovery
  - transparency/reuse across teams
  - self service
- lessons learned:
  - declarative nature changed communication between ops and devs
  - RBAC to git becomes your access control to infra, so it becomes more critical
  - shines a light on how you do configuration management
  - take time up front to consider who needs to see what and which UX the tools provide, how do you want to transition between environments
  - leads to looking hard at bits of infra that don't work declaritively
  - less spreading of credentials to the infra (flux sits inside the infra; one team runs flux on its own locked down cluster)
  - invest in alerting as automation doesn't guarantee its working
  - tooling for cross cluster routing is getting to where cluster can be made even more easily disposable
- sidenotes:
  - Tanka, Jsonnet emerging as alternatives to yaml
  - the speakers had barely any devs using gitops for app deployment!
    They are exited about the work on Cloud Native Application Bundle, given that things like schema migrations are not representable currently using declarative gitops.

### [Progressive Delivery Techniques with Flagger - Stefan Prodan, Weaveworks](https://www.youtube.com/watch?v=ahLnVkLlQ4U)

Good overview on how to split CI from CD. (Flagger is what we use for Reviewer)

- which deployment strat is suitable for which application
- flux v2 can be told about dependencies between things managed by it

### [Serverless or Servicefull - Sebastien Goasguen, Triggermesh](https://www.youtube.com/watch?v=g1A5bTRmuPI)

Shows nicely how the k8s API-CRDs-operator model is being lifted to _serverless_ world

- AWS, GCP, Azure are all working to letting people manage infra as CRDs
- connecting events between cloud services problem domain under active development
- _serverless_ is used to refer to a lot more that Lambda etc. (think S3, Cognito, CloudRun)

### [Cloud Native Live: Crossplane - GitOps-based Infrastructure as Code through Kubernetes API](https://www.youtube.com/watch?v=LyN50Ul2KOM)

Very interesting cross cloud provider IaC tool

- Crossplane hit v1.0 in December 2020
- GitOps-Operator approach to IaC (unlike Terraform which more in the pipeline model)

### [Keynote: Kubernetes Serverless Present and Serverless Future – Brendan Burns](https://www.youtube.com/watch?v=VfZ8y-4V6Yc)

Interesting: nodeless k8s and how it relates to IoT and traditional k8s

- Microsoft employee
- k8s API tries to be application focused but still knows about machines
- different meanings for _serverless_
  - FaaS
  - run containers without knowing about servers (hide the machines)
- Helm as example for providing a _programming model_ by layering on the k8s api
- how to hide the nodes when using k8s api?
  - parts that are machine oriented: scheduling, affinity
  - most other parts aren't concerned about machines
- challenges:
  - provide similar experience for k8s deploys regardless of whether they are deployed to _nodelss_ or traditional cluster
  - leverage e.g. rollout and deployment features for updating IoT devices
  - allowing nodeless to address non-linux nodes
  - how to surface constraints e.g. geographic region, availability zones
- Azure has _virtual kubelets_

### [Keynote: The Challenges of Funding Open Source Development - Todd Moore](https://www.youtube.com/watch?v=3C3mS9Xfb3w)

Nice historical perspective, but not much meat on the bone.

- IBM employee
- devs energized by _Peace, Love and Linux_
- Apache being OS becoming de facto standard
- groups of companies forming OS initiatives (Openstack etc.)
  _don't compete on plumbing_

### [Deep Dive: Falco - Jorge Salamero Sanz, Sysdig](https://www.youtube.com/watch?v=VfZ8y-4V6Yc)

Nice to know it exists. Feels too complex to be of value to us.

- developed by Sysdig in 2016
- behavioural security detection
- match syscall events against rules
- does the same for k8s audit events
- examples:
  - changes to configmap that introduces fields that indicate something that should be a secret
  - running an kubectl exec against a prod deployment

### [Securing the Deploy Pipeline - Felix Glaser, Shopify](https://www.youtube.com/watch?v=CZi9zV7OcyM)

Seems useful for orgs with lots of containers. Stack is designed to allow you to build very custom policies. General purpose tools like [Habor](https://goharbor.io/) might be sufficient.

I think autodeploying updates to the base image sha's could cover a lot of the same threat (wouldn't cover updates to deps installed in containers that aren't covered by e.g. package.json).

- mutability makes security harder
- threat model: inside the org someone deploys something vulnerable
- how to check if container is ok to run?
  - check before deploy at runtime? -> costs time
  - mark as ok once? -> vulns can appear after the check happened
- use of `admission controller` in k8s
- tools used: Kritis(policy enforcer), Grafeas(DB), Voucher(attestations)
- attestations managed by policies
  - defaults, per cluster...
  - types of vulns
  - break glass annotation
- workflow: Voucher called at build time of image to make attestations about it (doesn't run as root, free of known vulns..), on deploy Kritis looks up metadata in Grafeas which includes the data created by Voucher
- doesn't cover: continuous vuln scanning of deployed images

![image](https://user-images.githubusercontent.com/19282025/115722526-ff528a80-a376-11eb-8436-28f09c883425.png)

### 2021 CNCF End-user radar on Secret Manager

![image](https://user-images.githubusercontent.com/19282025/115727976-016b1800-a37c-11eb-9f34-1672913c2966.png)

### [Beyond File and Block Storage in Kubernetes - Sidhartha Mani, MinIO](https://www.youtube.com/watch?v=Y3GgJb71Cwo)

- working on bringing object storage to the k8s api akin to CSI

### [DevOps Performance From a Different Dataset: What 30M Workflows Reveal - Michael Stahnke, CircleCI](https://www.youtube.com/watch?v=eCS44PICvIk)

Dive into data from CircleCI builds (frequency, durations...)

- success rate is lower and duration higher for teams of one

### [Static Analysis of Kubernetes Manifests - Barak Schoster, Bridgecrew](https://www.youtube.com/watch?v=f0n8jPPrz6U)

`checkov` ist static analyzer for k8s, tf, cloudformation etc.

### [10 More Weird Ways to Blow Up Your Kubernetes - Jian Cheung & Joseph Kim, Airbnb](https://www.youtube.com/watch?v=4CT0cI62YHk)

- manual manipulation is dangerous
- find metrics that could have alerted you after incident
- there are interesting scaling breaking points in k8s (especially in controlplane) -> be ready to scale verticaclly
- CPU limits and throttling is complex
  - ran test that showed identical performance even though one was reporting being throttled
  - metrics aren't granular enough
  - percentage of throttled periods is better metric than absolute value of period
- cpu limit is per pod, not per container?

### [Making the Business Case for Contributing to Open Source - Tobie Langel, UnlockOpen](https://www.youtube.com/watch?v=p_Ig6sF1bHE)

Slick pitch for OS in orgs

- OpenSource the _complement_ to your offering
- leverage external contributions
- merge your changes upstream to remove cost of rebasing
- unrelated devs find each other via OS contributions
